core/block/object/objectcreator/systemobjectreviser.go

```diff
@@ -0,0 +1,160 @@
+package objectcreator
+
+import (
+	"strings"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/samber/lo"
+
+	"github.com/anyproto/anytype-heart/core/block/editor/basic"
+	"github.com/anyproto/anytype-heart/core/block/editor/smartblock"
+	"github.com/anyproto/anytype-heart/core/domain"
+	"github.com/anyproto/anytype-heart/core/relationutils"
+	"github.com/anyproto/anytype-heart/pb"
+	"github.com/anyproto/anytype-heart/pkg/lib/bundle"
+	"github.com/anyproto/anytype-heart/pkg/lib/database"
+	"github.com/anyproto/anytype-heart/pkg/lib/localstore/addr"
+	"github.com/anyproto/anytype-heart/pkg/lib/pb/model"
+	"github.com/anyproto/anytype-heart/space"
+	"github.com/anyproto/anytype-heart/util/pbtypes"
+)
+
+var revisionKey = bundle.RelationKeyRevision.String()
+
+func (s *service) reviseSystemObjects(space space.Space, objects map[string]*types.Struct) {
+	marketRels, err := s.objectStore.ListAllRelations(addr.AnytypeMarketplaceWorkspace)
+	if err != nil {
+		log.Errorf("failed to get relations from marketplace space: %v", err)
+		return
+	}
+
+	marketTypes, err := s.listAllObjectTypes(addr.AnytypeMarketplaceWorkspace)
+	if err != nil {
+		log.Errorf("failed to get object types from marketplace space: %v", err)
+		return
+	}
+
+	for source, details := range objects {
+		if strings.HasPrefix(source, addr.BundledRelationURLPrefix) {
+			reviseSystemRelation(space, relationutils.RelationFromStruct(details), marketRels)
+		} else if strings.HasPrefix(source, addr.BundledObjectTypeURLPrefix) {
+			reviseSystemObjectType(space, details, marketTypes)
+		}
+	}
+}
+
+func (s *service) listAllObjectTypes(spaceId string) (map[string]*types.Struct, error) {
+	records, _, err := s.objectStore.Query(database.Query{
+		Filters: []*model.BlockContentDataviewFilter{
+			{
+				RelationKey: bundle.RelationKeyLayout.String(),
+				Condition:   model.BlockContentDataviewFilter_Equal,
+				Value:       pbtypes.Float64(float64(model.ObjectType_objectType)),
+			},
+			{
+				RelationKey: bundle.RelationKeySpaceId.String(),
+				Condition:   model.BlockContentDataviewFilter_Equal,
+				Value:       pbtypes.String(spaceId),
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	details := make(map[string]*types.Struct, len(records))
+	for _, rec := range records {
+		id := pbtypes.GetString(rec.Details, bundle.RelationKeyId.String())
+		details[id] = rec.Details
+	}
+	return details, nil
+}
+
+func reviseSystemRelation(space space.Space, rel *relationutils.Relation, marketRels relationutils.Relations) {
```

Comment:
Maybe just use details structure? realationutils.Relation seems legacy to me

Reply:
Done!